### PR TITLE
Add dematerialize

### PR DESCRIPTION
**Pull request by paulyoung**
This pull request aims to enable going from `Decoded` to `throws` and vice-versa.

Currently this doesn't compile since the `catch` for `materialize` is not exhaustive and I'm not too sure what to do there.

### CODE REVIEW

**Comment by gfontenot** [comment]
Ha. You beat @tonyd256 to the PR. He was working on `dematerialize` last night.

I'm :+1: on `dematerialize`, but I'm not sure what the use case would be for `materialize`. I think that makes sense in Result, but I don't know if it applies here. Did you have ideas on when that would be useful?

Also, I think @tonyd256 was working on a test around `dematerialize`. Not sure if he got that working or not.

**Comment by paulyoung** [comment]
@gfontenot I think I agree with you on `materialize` but thought it might be nice to cover the edge case.

I only really added it because it was mentioned here: https://github.com/thoughtbot/Argo/pull/153#issuecomment-140873022

Tests for `dematerialize` on the way, unless @tonyd256 beats me to it.

**Comment by paulyoung** [comment]
Added some tests, please let me know if there is any feedback.

**Comment by tonyd256** [comment]
Damn you @paulyoung!! You beat me! haha ... anyways, This looks fine except I think you're going to run into the same issue I'm currently having.

Try changing you `catch` to catch the specific error the `try` should be throwing. I can't get that to work. ... Also feel free to look at my branch for what I mean.

File: Argo/Types/Decoded.swift [hunk 1]
```diff
@@ -76,3 +76,12 @@ public extension Decoded {
 public func pure<A>(a: A) -> Decoded<A> {
   return .Success(a)
 }
+
+public extension Decoded {
+  public func dematerialize() throws -> T {
```
Comment (tonyd256) [inline Argo/Types/Decoded.swift]: don't think you need public twice here

File: ArgoTests/Tests/DecodedTests.swift [hunk 2]
```diff
@@ -28,4 +28,26 @@ class DecodedTests: XCTestCase {
     default: XCTFail("Unexpected Case Occurred")
     }
   }
+  
+  func testDecodedDematerializeSuccess() {
+    let user: Decoded<User> = decode(JSONFromFile("user_with_email")!)
+    
+    do {
+      try user.dematerialize()
+      XCTAssertTrue(true, "Materialize threw an error with Decoded.Success")
```
Comment (tonyd256) [inline ArgoTests/Tests/DecodedTests.swift]: these error messages look off?

**Comment by paulyoung** [comment]
@tonyd256 just pushed some changes based on your comments (and your branch).

**Comment by tonyd256** [comment]
did the tests pass for you?

I finally got it to pass by adding `Argo.` to the front of those cases ... see updates to my branch

**Comment by paulyoung** [comment]
@tonyd256 everything seems to work as expected. Can you confirm?

File: ArgoTests/Tests/DecodedTests.swift [hunk 3]
```diff
@@ -28,4 +28,45 @@ class DecodedTests: XCTestCase {
...
+    }
+  }
+  
+  func testDecodedDematerializeTypeMismatch() {
+    let user: Decoded<User> = decode(JSONFromFile("user_with_bad_type")!)
+    
+    do {
+      try user.dematerialize()
+      XCTFail("Unexpected Success")
+    } catch DecodeError.TypeMismatch(_, _) {
```
Comment (tonyd256) [inline ArgoTests/Tests/DecodedTests.swift]: looks like we can just leave off all the `(_)` and `(_,_)`s in all these

Comment (paulyoung) [inline ArgoTests/Tests/DecodedTests.swift]: Done.

File: ArgoTests/Tests/DecodedTests.swift [hunk 4]
```diff
@@ -28,4 +28,45 @@ class DecodedTests: XCTestCase {
...
+  }
+  
+  func testDecodedDematerializeTypeMismatch() {
+    let user: Decoded<User> = decode(JSONFromFile("user_with_bad_type")!)
+    
+    do {
+      try user.dematerialize()
+      XCTFail("Unexpected Success")
+    } catch DecodeError.TypeMismatch {
+      XCTAssertTrue(true)
```
Comment (tonyd256) [inline ArgoTests/Tests/DecodedTests.swift]: what to shorten this to `XCTAssert(true)` and the one below

**Comment by tonyd256** [comment]
@paulyoung on small change and then could you squash this down to one commit?

File: ArgoTests/Tests/DecodedTests.swift (hunk 4 above)
Comment (paulyoung) [inline ArgoTests/Tests/DecodedTests.swift]: Thanks. Thought I'd got all of those.

**Comment by paulyoung** [comment]
Squashed :eggplant:

**Comment by tonyd256** [comment]
:boom: :muscle: :us: .... Thanks @paulyoung
